src/sas/sascalc/calculator/geni.py

```diff
@@ -514,4 +516,109 @@ def _spin_weights(in_spin, out_spin):
+
+    rog_num = np.sum(masses * rsq)
+    rog_den = np.sum(masses)
+    guinier_num = np.sum(contrastSLDs * rsq)
+    guinier_den = np.sum(contrastSLDs)
+
+
+    if rog_den <= 0: #Should never happen as there are no zero or negative mass atoms
+        rog_mass = "NaN"
+        logging.warning("Atomic Mass is zero for all atoms in the system.")
+    else:
+        rGMass = np.sqrt(rog_num/rog_den)
+        rog_mass = (str(round(np.sqrt(rog_num/rog_den),1)) + " Å")
+
+    #Avoid division by zero - May occur through contrast matching
+    if guinier_den == 0:
+        guinier_value = "NaN"
+        logging.warning("Effective Coherent Scattering Length is zero for all atoms in the system.")
+    elif (guinier_num/guinier_den) < 0:
+        guinier_value = (str(round(np.sqrt(-guinier_num/guinier_den),1)) + " Å")
+        logging.warning("Radius of Gyration Squared is negative. R(G)^2 is assumed to be |R(G)|* R(G).")
+    else:
+        guinier_value = (str(round(np.sqrt(guinier_num/guinier_den),1)) + " Å")
+
+    return [rog_mass,guinier_value,rGMass]          #[String, String, Float], float used for plugin model
+
+    
+def centerOfMass(nuc_sld_data):
+    """Calculate Center of Mass(CoM) of provided atom"""
+    CoMnumerator= [0.0,0.0,0.0]
+    CoMdenominator = [0.0,0.0,0.0]
+
+    for i in range(len(nuc_sld_data.pos_x)):
+        coordinates = [float(nuc_sld_data.pos_x[i]),float(nuc_sld_data.pos_y[i]),float(nuc_sld_data.pos_z[i])]
+        
+        #Coh b - Coherent Scattering Length(fm)
+        cohB = periodictable.elements.symbol(nuc_sld_data.pix_symbol[i]).neutron.b_c
+
+        for j in range(3): #sets CiN
+            CoMnumerator[j] += (coordinates[j]*cohB)
+            CoMdenominator[j] += cohB
+
+    CoM = [] 
+    for i in range(3):   
+        CoM.append(CoMnumerator[i]/CoMdenominator[i] if CoMdenominator != 0 else 0) #center of mass, test for division by zero
+    
+    return CoM
+    
+def create_betaPlot(qX, nuc_sld_data, npts_x, formFactor):
+    """Carry out the compuation of beta Q using provided & calculated data
+    Returns a list of BetaQ values
+
+    """
+    fQ = FQ(qX, nuc_sld_data, npts_x)
+    
+    #Center Of Mass Calculation
+    data_betaQ = (fQ**2)/formFactor
+
+    
+    #Scale Beta Q to 0-1
+    scalingFactor = data_betaQ[0]
+    data_betaQ = data_betaQ / scalingFactor
+
+    return data_betaQ
+
+
+def FQ(qX, nuc_sld_data, npts_x):
+    fQlist = np.empty(npts_x)
+    CoM = centerOfMass(nuc_sld_data)
+    r_x = np.subtract(nuc_sld_data.pos_x , CoM[0])
+    r_y = np.subtract(nuc_sld_data.pos_y , CoM[1])
+    r_z = np.subtract(nuc_sld_data.pos_z , CoM[2])
+    magnitudeRelativeCoordinate = np.sqrt(np.power(r_x, 2) + np.power(r_y, 2) + np.power(r_z, 2))
+    cohB = np.asarray([periodictable.elements.symbol(atom).neutron.b_c for atom in nuc_sld_data.pix_symbol])
+    
+    for i in range(npts_x):
+        fQ = np.sum(cohB*(np.sin(qX[i] * magnitudeRelativeCoordinate) / (qX[i] * magnitudeRelativeCoordinate)))
```

Comment:
Fails if qX[i] = 0 or any r = center of mass.

You can use `np.sinc((qX[i]/np.pi) * dr)` to handle sin(0)/0.

This runs a bit slower than the incorrect sin(qr)/qr on my system. If it is a concern you can wrap it in numba.njit.

The following requires 50 μs on my machine vs 70 μs for Σ b sin(qΔr)/qΔr) and 85 μs for Σ b sinc(qΔr/π))
```python
@numba.njit
def f(q, b, dr):
     total = 0
     for bi, dri in zip(b,dr):
         qr = q*dri
         total += bi * np.sin(qr)/qr if qr != 0. else bi
     return total
```

Reply:
Has this been resolved?

Reply:
No, also I agree

Reply:
np.sin(qr)/qr is changed to np.sinc(qr/np.pi).